pyo3-macros-backend/src/pyclass.rs

```diff
@@ -640,6 +768,157 @@ fn impl_enum(
+            deprecations: Deprecations::new(),
+        };
+        let variant_new = complex_enum_variant_new(cls, variant)?;
+
+        let (variant_cls_impl, field_getters) = impl_complex_enum_variant_cls(cls, &variant)?;
+        variant_cls_impls.push(variant_cls_impl);
+
+        let pyclass_impl = PyClassImplsBuilder::new(
+            &variant_cls,
+            &variant_args,
+            PyClassMethodsType::Specialization,
+            field_getters,
+            vec![variant_new],
+        )
+        .impl_all()?;
+
+        variant_cls_pyclass_impls.push(pyclass_impl);
+    }
+
+    Ok(quote! {
+        const _: () = {
+            use #krate as _pyo3;
+
+            #pytypeinfo
+
+            #pyclass_impls
+
+            #[doc(hidden)]
+            #[allow(non_snake_case)]
+            impl #cls {}
+
+            #(#variant_cls_newtypes)*
+
+            #(#variant_cls_pytypeinfos)*
+
+            #(#variant_cls_pyclass_impls)*
+
+            #(#variant_cls_impls)*
+        };
+    })
+}
+
+fn impl_complex_enum_variant_cls(
+    enum_name: &syn::Ident,
+    variant: &PyClassComplexEnumVariant<'_>,
+) -> Result<(TokenStream, Vec<MethodAndMethodDef>)> {
+    let variant_ident = &variant.ident;
+    let variant_cls = gen_complex_enum_variant_class_ident(enum_name, &variant.ident);
+    let variant_cls_type = parse_quote!(#variant_cls);
+
+    let mut field_names: Vec<Ident> = vec![];
+    let mut fields_with_types: Vec<TokenStream> = vec![];
+    let mut field_getters = vec![];
+    let mut field_getter_impls: Vec<TokenStream> = vec![];
+    for field in &variant.fields {
+        let field_name = field.ident;
+        let field_type = field.ty;
+        let field_with_type = quote! { #field_name: #field_type };
+
+        let field_getter = complex_enum_variant_field_getter(
+            &variant_cls_type,
+            &field_name,
+            field_type,
+            field.span,
+        )?;
+
+        let field_getter_impl = quote! {
+            fn #field_name(&self) -> _pyo3::PyResult<#field_type> {
+                match &self.0 {
+                    // TODO(mkovaxx): Fix this to work with tuple variants
+                    #enum_name::#variant_ident { #field_name, .. } => Ok(#field_name.clone()),
+                    _ => unreachable!("Wrong complex enum variant found in wrapper newtype"),
+                }
+            }
+        };
+
+        field_names.push(field_name.clone());
+        fields_with_types.push(field_with_type);
+        field_getters.push(field_getter);
+        field_getter_impls.push(field_getter_impl);
+    }
+
+    let cls_impl = quote! {
+        #[doc(hidden)]
+        #[allow(non_snake_case)]
+        impl #variant_cls {
+            fn __pymethod_constructor__(#(#fields_with_types)*) -> Self {
+                // TODO(mkovaxx): Fix this to work with tuple variants
+                Self(#enum_name::#variant_ident { #(#field_names)* })
+            }
```

Comment:
With the zero-sized variant types I think this will need to be something more like:

```suggestion
            fn __pymethod_constructor__(py: Python<'_>, #(#fields_with_types)*) -> PyObject {
                // TODO(mkovaxx): Fix this to work with tuple variants
                #enum_name::#variant_ident { #(#field_names)* }.into_py(py)
            }
```

... and we can lean on the `IntoPy` implementation working correctly.